tests/unit/s2n_random_test.c

```diff
@@ -184,169 +108,376 @@ int main(int argc, char **argv)
+    EXPECT_OK(s2n_thread_test(s2n_get_public_random_data, s2n_get_private_random_data));
+    EXPECT_OK(s2n_thread_test(s2n_get_private_random_data, s2n_get_public_random_data));
+
+    /* Verify we generate unique data over forks */
+    EXPECT_OK(s2n_fork_test(s2n_get_private_random_data, s2n_get_private_random_data));
+    EXPECT_OK(s2n_fork_test(s2n_get_public_random_data, s2n_get_public_random_data));
+    EXPECT_OK(s2n_fork_test(s2n_get_public_random_data, s2n_get_private_random_data));
+    EXPECT_OK(s2n_fork_test(s2n_get_private_random_data, s2n_get_public_random_data));
+
+    /* Basic tests generating randomness */
+    EXPECT_OK(s2n_basic_generate_tests());
+
+    /* Verify that there are no trivially observable patterns in the output */
+    EXPECT_OK(s2n_basic_pattern_tests(s2n_get_public_random_data));
+    EXPECT_OK(s2n_basic_pattern_tests(s2n_get_private_random_data));
+
+    /* Just a sanity check and avoids cppcheck "unassignedVariable" errors.
+     * In future PRs this part will be expanded.
+     */
+    blob1.size = RANDOM_GENERATE_DATA_SIZE;
+    EXPECT_OK(s2n_get_public_random_data(&blob1));
+    blob2.size = RANDOM_GENERATE_DATA_SIZE;
+    EXPECT_OK(s2n_get_private_random_data(&blob2));
+    EXPECT_BYTEARRAY_NOT_EQUAL(data1, data2, RANDOM_GENERATE_DATA_SIZE);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_random_test_case_default_cb(struct random_test_case *test_case)
+{
+    EXPECT_SUCCESS(s2n_init());
+
+    /* Verify that randomness callbacks can't be set to NULL */
+    EXPECT_FAILURE(s2n_rand_set_callbacks(NULL, s2n_cleanup_cb, s2n_entropy_cb, s2n_entropy_cb));
+    EXPECT_FAILURE(s2n_rand_set_callbacks(s2n_init_cb, NULL, s2n_entropy_cb, s2n_entropy_cb));
+    EXPECT_FAILURE(s2n_rand_set_callbacks(s2n_init_cb, s2n_cleanup_cb, NULL, s2n_entropy_cb));
+    EXPECT_FAILURE(s2n_rand_set_callbacks(s2n_init_cb, s2n_cleanup_cb, s2n_entropy_cb, NULL));
+
+    EXPECT_EQUAL(s2n_common_tests(test_case), S2N_SUCCESS);
+
+    EXPECT_SUCCESS(s2n_cleanup());
+
+    return EXIT_SUCCESS;
+}
+
+static int s2n_random_test_case_without_pr_cb(struct random_test_case *test_case)
+{
+    EXPECT_SUCCESS(s2n_init());
+
+    POSIX_GUARD_RESULT(s2n_ignore_prediction_resistance_for_testing(true));
+    EXPECT_EQUAL(s2n_common_tests(test_case), S2N_SUCCESS);
+    POSIX_GUARD_RESULT(s2n_ignore_prediction_resistance_for_testing(false));
+
+    EXPECT_SUCCESS(s2n_cleanup());
+
+    return EXIT_SUCCESS;
+}
+
+static int s2n_random_test_case_failure_cb(struct random_test_case *test_case)
+{
+    EXPECT_SUCCESS(s2n_init());
+
+    /* This is a cheap way to ensure that failures in a fork bubble up to the
+     * parent as a failure. This should be caught in the parent when querying
+     * the return status code of the child. All test macro's will cause a
+     * process to exit with EXIT_FAILURE. We call exit() directly to avoid
+     * messages being printed on stderr, in turn, appearing in logs.
+     */
+    exit(1);
```

Comment:
Maybe something like EXPECT_TRUE(false) to fail in the same way the tests will? Then you don't have to rely on the test macros behaving a specific way.